README.md
[![Build Status](https://travis-ci.org/tumugi/tumugi-plugin-google_cloud_storage.svg?branch=master)](https://travis-ci.org/tumugi/tumugi-plugin-google_cloud_storage) [![Code Climate](https://codeclimate.com/github/tumugi/tumugi-plugin-google_cloud_storage/badges/gpa.svg)](https://codeclimate.com/github/tumugi/tumugi-plugin-google_cloud_storage) [![Coverage Status](https://coveralls.io/repos/github/tumugi/tumugi-plugin-google_cloud_storage/badge.svg?branch=master)](https://coveralls.io/github/tumugi/tumugi-plugin-google_cloud_storage?branch=master) [![Gem Version](https://badge.fury.io/rb/tumugi-plugin-google_cloud_storage.svg)](https://badge.fury.io/rb/tumugi-plugin-google_cloud_storage)

# Google Cloud Storage plugin for [tumugi]((https://github.com/tumugi/tumugi)

tumugi-plugin-google_cloud_storage is a plugin for integrate [Google Cloud Storage](https://cloud.google.com/storage/) and [tumugi](https://github.com/tumugi/tumugi).

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'tumugi-plugin-google_cloud_storage'
```

And then execute `bundle install`

## Target

### Tumugi::Plugin::GoogleCloudStorageFileTarget

This target represent file or directory of Googl Cloud Storage.

#### Paramters

| name   | type   | required? | default | description                                                                              |
|--------|--------|-----------|---------|------------------------------------------------------------------------------------------|
| bucket | string | required  |         | [bucket](https://cloud.google.com/storage/docs/json_api/v1/buckets) name of GCS          |
| key    | string | required  |         | key (= [object](https://cloud.google.com/storage/docs/json_api/v1/objects) name) of GCS. |

#### Examples

Create a file which content is "done" in Google Cloud Storage.

```rb
task :task1 do
  param :bucket, type: :string, auto_bind: true, required: true
  param :day, type: :time, auto_bind: true, required: true

  output do
    target(:google_cloud_storage_file,
      bucket: bucket,
      key: "test_#{day.strftime('%Y%m%d')}.txt")
  end

  run do
    log 'task1#run'
    output.open('w') {|f| f.puts('done') }
  end
end
```

Execute this file:

```sh
$ bundle exec tumugi run -f workflow.rb -p bucket:BUCKET_NAME day:2016-07-01
```

### Config Section

tumugi-plugin-google_cloud_storage provide config section named "google_cloud_storage" which can specified Google Cloud Storage autenticaion info.

#### Authenticate by client_email and private_key

```rb
Tumugi.config do |config|
  config.section("google_cloud_storage") do |section|
    section.project_id = "xxx"
    section.client_email = "yyy@yyy.iam.gserviceaccount.com"
    section.private_key = "zzz"
  end
end
```

#### Authenticate by JSON key file

```rb
Tumugi.configure do |config|
  config.section("google_cloud_storage") do |section|
    section.private_key_file = "/path/to/key.json"
  end
end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/tumugi/ttumugi-plugin-google_cloud_storage

## License

The gem is available as open source under the terms of the [Apache License
Version 2.0](http://www.apache.org/licenses/).


CHANGELOG.md
# Change Log

## [v0.3.0](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/tree/v0.3.0) (2016-07-18)
[Full Changelog](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/compare/v0.2.0...v0.3.0)

**Merged pull requests:**

- Update README [\#19](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/19) ([hakobera](https://github.com/hakobera))
- Update tumugi to 0.6 [\#18](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/18) ([hakobera](https://github.com/hakobera))

## [v0.2.0](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/tree/v0.2.0) (2016-06-06)
[Full Changelog](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/compare/v0.1.0...v0.2.0)

**Implemented enhancements:**

- Change tumugi dependency version [\#14](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/14) ([hakobera](https://github.com/hakobera))
- Support custom filesystem for GoogleCloudStorageFileTarget [\#11](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/11) ([hakobera](https://github.com/hakobera))

**Merged pull requests:**

- Update changelog [\#17](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/17) ([hakobera](https://github.com/hakobera))
- Remove unneeded log [\#16](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/16) ([hakobera](https://github.com/hakobera))
- Prepare release for 0.2.0 [\#15](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/15) ([hakobera](https://github.com/hakobera))
- Change test directory [\#13](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/13) ([hakobera](https://github.com/hakobera))
- Change ruby version on travis [\#12](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/12) ([hakobera](https://github.com/hakobera))

## [v0.1.0](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/tree/v0.1.0) (2016-05-30)
**Closed issues:**

- Release v0.1.0 [\#8](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/issues/8)

**Merged pull requests:**

- Preapre release for v0.1.0 [\#10](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/10) ([hakobera](https://github.com/hakobera))
- Support JSON private key file for authentication [\#9](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/9) ([hakobera](https://github.com/hakobera))
- Update README [\#7](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/7) ([hakobera](https://github.com/hakobera))
- Standardize Google API Call error handling [\#6](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/6) ([hakobera](https://github.com/hakobera))
- Fix railse\_if\_exist option not working when copy and move [\#5](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/5) ([hakobera](https://github.com/hakobera))
- Rename gem and namespace [\#4](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/4) ([hakobera](https://github.com/hakobera))
- Add tests [\#3](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/3) ([hakobera](https://github.com/hakobera))
- Update tumugi to v0.5.1 [\#2](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/2) ([hakobera](https://github.com/hakobera))
- Fix test failed on travis [\#1](https://github.com/tumugi/tumugi-plugin-google_cloud_storage/pull/1) ([hakobera](https://github.com/hakobera))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

test/test_helper.rb
require 'coveralls'
Coveralls.wear!

$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

require 'test/unit'
require 'test/unit/rr'

require 'tumugi'

Dir.mkdir('tmp') unless Dir.exist?('tmp')

def credential
  pkey = ENV['PRIVATE_KEY'].gsub(/\\n/, "\n")
  OpenStruct.new({
    project_id: ENV['PROJECT_ID'],
    client_email: ENV['CLIENT_EMAIL'],
    private_key: pkey
  })
end

def credential_file
  file = 'tmp/credential.json'
  pkey = ENV['PRIVATE_KEY'].gsub(/\\n/, "\n")
  File.write(file, JSON.generate({
    project_id: ENV['PROJECT_ID'],
    client_email: ENV['CLIENT_EMAIL'],
    private_key: pkey
  }))
  OpenStruct.new(private_key_file: file)
end

Tumugi.configure do |config|
  config.section('google_cloud_storage') do |section|
    section.project_id = credential[:project_id]
    section.client_email = credential[:client_email]
    section.private_key = credential[:private_key]
  end
end


test/tumugi/plugin/gcs/atomic_gcs_file_test.rb
require_relative '../../../test_helper'
require 'tumugi/plugin/gcs/atomic_gcs_file'
require 'tumugi/plugin/gcs/gcs_file_system'

class Tumugi::Plugin::GCS::AtomicGCSFileTest < Test::Unit::TestCase
  setup do
    @fs = Tumugi::Plugin::GCS::GCSFileSystem.new(credential)
    @bucket = "tumugi-plugin-gcs"
    @prefix = "#{SecureRandom.hex(10)}"
  end

  teardown do
    @fs.remove("gs://#{@bucket}/#{@prefix}/")
  end

  test "after open and close file, file upload to GCS" do
    path = "gs://#{@bucket}/#{@prefix}/atomic_gcs_file_test.txt"
    @file = Tumugi::Plugin::GCS::AtomicGCSFile.new(path, @fs)
    @file.open do |f|
      f.puts 'test'
    end
    @fs.exist?(path)
    @fs.download(path) do |f|
      assert_equal("test\n", f.read)
    end
  end
end


test/tumugi/plugin/gcs/gcs_file_system_test.rb
require_relative '../../../test_helper'
require 'tumugi/plugin/gcs/gcs_file_system'

class Tumugi::Plugin::GCS::GCSFileSystemTest < Test::Unit::TestCase
  setup do
    @fs = Tumugi::Plugin::GCS::GCSFileSystem.new(credential)
    @bucket = "tumugi-plugin-gcs"
    @prefix = "#{SecureRandom.hex(10)}"
    @keys = [ "file1.txt", "folder/file2.txt"]
    @keys.each do |key|
      @fs.put_string('test', "gs://#{@bucket}/#{@prefix}/fs_test/#{key}")
    end
  end

  teardown do
    @fs.remove("gs://#{@bucket}/#{@prefix}/")
  end

  sub_test_case "exist?" do
    test "true" do
      assert_true(@fs.exist?("gs://#{@bucket}/#{@prefix}/fs_test/"))
      assert_true(@fs.exist?("gs://#{@bucket}/#{@prefix}/fs_test/#{@keys[0]}"))
    end

    test "false" do
      assert_false(@fs.exist?("gs://#{@bucket}/#{@prefix}/fs_test/not_found_dir/"))
      assert_false(@fs.exist?("gs://#{@bucket}/#{@prefix}/fs_test/not_found_file.txt"))
    end
  end

  sub_test_case "remove" do
    test "cannot delete root" do
      assert_raise(Tumugi::FileSystemError) do
        @fs.remove("gs://#{@bucket}")
      end
    end

    test "exist file" do
      assert_true(@fs.remove("gs://#{@bucket}/#{@prefix}/fs_test/#{@keys[0]}"))
    end

    test "non exist file" do
      assert_false(@fs.remove("gs://#{@bucket}/#{@prefix}/fs_test/not_found.txt"))
    end

    test "directory without recursive flag" do
      assert_raise(Tumugi::FileSystemError) do
        @fs.remove("gs://#{@bucket}/#{@prefix}/fs_test/", recursive: false)
      end
    end

    test "directory also delete child files" do
      assert_true(@fs.remove("gs://#{@bucket}/#{@prefix}/fs_test/"))
      assert_equal(0, @fs.entries("gs://#{@bucket}/#{@prefix}/fs_test/").count)
    end
  end

  sub_test_case "mkdir" do
    test "return true when success" do
      assert_true(@fs.mkdir("gs://#{@bucket}/#{@prefix}/fs_test/new_dir/"))
    end

    sub_test_case "path is already exist" do
      test "raise error if raise_if_exist flag is true" do
        assert_raise(Tumugi::FileAlreadyExistError) do
          @fs.mkdir("gs://#{@bucket}/#{@prefix}/fs_test/", raise_if_exist: true)
        end
      end

      test "raise error if path is not a directory" do
        assert_raise(Tumugi::NotADirectoryError) do
          @fs.mkdir("gs://#{@bucket}/#{@prefix}/fs_test/#{@keys[0]}")
        end
      end

      test "return false" do
        assert_false(@fs.mkdir("gs://#{@bucket}/#{@prefix}/fs_test/"))
      end
    end
  end

  data({
    "root" => [ true, "" ],
    "exist" => [ true, "fs_test/" ],
    "prefix" => [ true, "fs_test/folder" ],
    "not_a_folder" => [ false, "fs_test/not_a_folder" ]
  })
  test "directory?" do |(expected, path)|
    assert_equal(expected, @fs.directory?("gs://#{@bucket}/#{@prefix}/#{path}"))
  end

  sub_test_case "entries" do
    test "path has entries" do
      entries = @fs.entries("gs://#{@bucket}/#{@prefix}/fs_test/")
      assert_equal(@keys.count, entries.count)
      @keys.each_with_index do |key, i|
        assert_equal(@bucket, entries[i].bucket)
        assert_equal("#{@prefix}/fs_test/#{key}", entries[i].name)
      end
    end

    test "path has no entry" do
      entries = @fs.entries("gs://#{@bucket}/#{@prefix}/fs_test/no_entries/")
      assert_true(entries.empty?)
    end
  end

  sub_test_case "move" do
    test "directory" do
      src_path = "gs://#{@bucket}/#{@prefix}/fs_test/"
      dest_path = "gs://#{@bucket}/#{@prefix}/fs_dest/"

      @fs.move(src_path, dest_path)

      assert_false(@fs.exist?(src_path))
      entries = @fs.entries(dest_path)
      assert_equal(@keys.count, entries.count)
      @keys.each_with_index do |key, i|
        assert_equal(@bucket, entries[i].bucket)
        assert_equal("#{@prefix}/fs_dest/#{key}", entries[i].name)
      end
    end
  end

  test "upload" do
    new_file_path = "gs://#{@bucket}/#{@prefix}/fs_test/new_file.txt"
    @fs.upload(StringIO.new('upload'), new_file_path)
    assert_true(@fs.exist?(new_file_path))
  end

  sub_test_case "download" do
    test "without block" do
      f = @fs.download("gs://#{@bucket}/#{@prefix}/fs_test/#{@keys[0]}", download_path: 'tmp/download.txt')
      assert_true(File.exist?('tmp/download.txt'))
      assert_equal('test', f.read)
      f.close
    end

    test "with block, without destination file name" do
      @fs.download("gs://#{@bucket}/#{@prefix}/fs_test/#{@keys[0]}") do |f|
        assert_equal('test', f.read)
      end
    end
  end

  sub_test_case "copy" do
    test "directory" do
      src_path = "gs://#{@bucket}/#{@prefix}/fs_test/"
      dest_path = "gs://#{@bucket}/#{@prefix}/fs_dest/"

      @fs.copy(src_path, dest_path)

      entries = @fs.entries(dest_path)
      assert_equal(@keys.count, entries.count)
      @keys.each_with_index do |key, i|
        assert_equal(@bucket, entries[i].bucket)
        assert_equal("#{@prefix}/fs_dest/#{key}", entries[i].name)
      end
    end
  end

  sub_test_case "path_to_bucket_and_key" do
    test "success" do
      result = @fs.path_to_bucket_and_key("gs://bucket/path/to/object")
      assert_equal('bucket', result[0])
      assert_equal('path/to/object', result[1])
    end

    test "raise error when scheme is not 'gs'" do
      assert_raise(Tumugi::FileSystemError) do
        @fs.path_to_bucket_and_key("https://bucket/path/to/object")
      end
    end
  end

  test "bucket releated methods" do
    bucket = "tumugi-plugin-gcs-#{SecureRandom.hex(10)}"
    assert_false(@fs.bucket_exist?(bucket))
    assert_true(@fs.create_bucket(bucket))
    assert_true(@fs.bucket_exist?(bucket))
    assert_true(@fs.remove_bucket(bucket))
    assert_false(@fs.bucket_exist?(bucket))
  end
end


lib/tumugi/plugin/gcs/gcs_file_system.rb
require 'uri'
require 'google/apis/storage_v1'
require 'tumugi/file_system'

module Tumugi
  module Plugin
    module GCS
      class GCSFileSystem < Tumugi::FileSystem
        attr_reader :client

        def initialize(config)
          @project_id = config[:project_id] || config.project_id
          @client = create_client(config)
        end

        #######################################################################
        # FileSystem interfaces
        #######################################################################

        def exist?(path)
          bucket, key = path_to_bucket_and_key(path)
          if obj_exist?(bucket, key)
            true
          else
            directory?(path)
          end
        end

        def remove(path, recursive: true)
          bucket, key = path_to_bucket_and_key(path)
          raise Tumugi::FileSystemError.new("Cannot delete root of bucket at path '#{path}'") if root?(key)

          if obj_exist?(bucket, key)
            @client.delete_object(bucket, key)
            wait_until { !obj_exist?(bucket, key) }
            true
          elsif directory?(path)
            raise Tumugi::FileSystemError.new("Path '#{path}' is a directory. Must use recursive delete") if !recursive

            objs = entries(path).map(&:name)
            @client.batch do |client|
              objs.each do |obj|
                client.delete_object(bucket, obj)
              end
            end
            wait_until { !directory?(path) }
            true
          else
            false
          end
        end

        def mkdir(path, parents: true, raise_if_exist: false)
          if exist?(path)
            if raise_if_exist
              raise Tumugi::FileAlreadyExistError.new("Path #{path} is already exist")
            elsif !directory?(path)
              raise Tumugi::NotADirectoryError.new("Path #{path} is not a directory")
            end
            false
          else
            put_string("", add_path_delimiter(path))
            true
          end
        end

        def directory?(path)
          bucket, key = path_to_bucket_and_key(path)
          if root?(key)
            bucket_exist?(bucket)
          else
            obj = add_path_delimiter(key)
            if obj_exist?(bucket, obj)
              true
            else
              # Any objects with this prefix
              objects = @client.list_objects(bucket, prefix: obj, max_results: 20)
              !!(objects.items && objects.items.size > 0)
            end
          end
        end

        def entries(path)
          bucket, key = path_to_bucket_and_key(path)
          obj = add_path_delimiter(key)
          results = []
          next_page_token = ''

          until next_page_token.nil?
            objects = @client.list_objects(bucket, prefix: obj, page_token: next_page_token)
            if objects && objects.items
              results.concat(objects.items)
              next_page_token = objects.next_page_token
            else
              next_page_token = nil
            end
          end
          results
        end

        def move(src_path, dest_path, raise_if_exist: false)
          copy(src_path, dest_path, raise_if_exist: raise_if_exist)
          remove(src_path)
        end

        #######################################################################
        # Specific methods
        #######################################################################

        def upload(media, path, content_type: nil)
          bucket, key = path_to_bucket_and_key(path)
          obj = Google::Apis::StorageV1::Object.new(bucket: bucket, name: key)
          @client.insert_object(bucket, obj, upload_source: media, content_type: content_type)
          wait_until { obj_exist?(bucket, key) }
        end

        def download(path, download_path: nil, mode: 'r', &block)
          bucket, key = path_to_bucket_and_key(path)
          if download_path.nil?
            download_path = Tempfile.new('tumugi_gcs_file_system').path
          end
          @client.get_object(bucket, key, download_dest: download_path)
          wait_until { File.exist?(download_path) }

          if block_given?
            File.open(download_path, mode, &block)
          else
            File.open(download_path, mode)
          end
        end

        def put_string(contents, path, content_type: 'text/plain')
          media = StringIO.new(contents)
          upload(media, path, content_type: content_type)
        end

        def copy(src_path, dest_path, raise_if_exist: false)
          src_bucket, src_key = path_to_bucket_and_key(src_path)
          dest_bucket, dest_key = path_to_bucket_and_key(dest_path)

          if directory?(src_path)
            src_prefix = add_path_delimiter(src_key)
            dest_prefix = add_path_delimiter(dest_key)

            src_path = add_path_delimiter(src_path)
            copied_objs = []
            entries(src_path).each do |entry|
              suffix = entry.name[src_prefix.length..-1]
              @client.copy_object(src_bucket, src_prefix + suffix,
                                  dest_bucket, dest_prefix + suffix)
              copied_objs << (dest_prefix + suffix)
            end
            wait_until { copied_objs.all? {|obj| obj_exist?(dest_bucket, obj)} }
          else
            @client.copy_object(src_bucket, src_key, dest_bucket, dest_key)
            wait_until { obj_exist?(dest_bucket, dest_key) }
          end
        end

        def path_to_bucket_and_key(path)
          uri = URI.parse(path)
          raise Tumugi::FileSystemError.new("URI scheme must be 'gs' but '#{uri.scheme}'") unless uri.scheme == 'gs'
          [ uri.host, uri.path[1..-1] ]
        end

        def create_bucket(bucket)
          unless bucket_exist?(bucket)
            b = Google::Apis::StorageV1::Bucket.new(name: bucket)
            @client.insert_bucket(@project_id, b)
            true
          else
            false
          end
        rescue => e
          raise Tumugi::FileSystemError.new(e.message)
        end

        def remove_bucket(bucket)
          if bucket_exist?(bucket)
            @client.delete_bucket(bucket)
            true
          else
            false
          end
        rescue => e
          raise Tumugi::FileSystemError.new(e.message)
        end

        def bucket_exist?(bucket)
          @client.get_bucket(bucket)
          true
        rescue => e
          return false if e.status_code == 404
          raise Tumugi::FileSystemError.new(e.message)
        end

        private

        def obj_exist?(bucket, key)
          @client.get_object(bucket, key)
          true
        rescue => e
          return false if e.status_code == 404
          raise Tumugi::FileSystemError.new(e.message)
        end

        def root?(key)
          key.nil? || key == ''
        end

        def add_path_delimiter(key)
          if key.end_with?('/')
            key
          else
            "#{key}/"
          end
        end

        def create_client(config)
          client_email = config[:client_email] || config.client_email
          private_key = config[:private_key] || config.private_key

          # https://cloud.google.com/storage/docs/authentication
          scope = "https://www.googleapis.com/auth/devstorage.read_write"

          if client_email and private_key
            auth = Signet::OAuth2::Client.new(
              token_credential_uri: "https://accounts.google.com/o/oauth2/token",
              audience: "https://accounts.google.com/o/oauth2/token",
              scope: scope,
              issuer: client_email,
              signing_key: OpenSSL::PKey.read(private_key))
            # MEMO: signet-0.6.1 depend on Farady.default_connection
            Faraday.default_connection.options.timeout = 60
            auth.fetch_access_token!
          else
            auth = Google::Auth.get_application_default([scope])
            auth.fetch_access_token!
          end

          client = Google::Apis::StorageV1::StorageService.new
          client.authorization = auth
          client
        end

        def wait_until(&block)
          while not block.call
            sleep 1
          end
        end
      end
    end
  end
end


lib/tumugi/plugin/google_cloud_storage/file_system.rb
require 'uri'
require 'json'
require 'googleauth/service_account'
require 'google/apis/storage_v1'
require 'google/apis/drive_v3'
require 'tumugi/file_system'

module Tumugi
  module Plugin
    module GoogleCloudStorage
      class FileSystem < Tumugi::FileSystem
        attr_reader :client

        def initialize(config)
          save_config(config)
        end

        #######################################################################
        # FileSystem interfaces
        #######################################################################

        def exist?(path)
          bucket, key = path_to_bucket_and_key(path)
          if obj_exist?(bucket, key)
            true
          else
            directory?(path)
          end
        rescue
          process_error($!)
        end

        def remove(path, recursive: true)
          bucket, key = path_to_bucket_and_key(path)
          raise Tumugi::FileSystemError.new("Cannot delete root of bucket at path '#{path}'") if root?(key)

          if obj_exist?(bucket, key)
            client.delete_object(bucket, key, options: request_options)
            wait_until { !obj_exist?(bucket, key) }
            true
          elsif directory?(path)
            raise Tumugi::FileSystemError.new("Path '#{path}' is a directory. Must use recursive delete") if !recursive

            objs = entries(path).map(&:name)
            client.batch do |client|
              objs.each do |obj|
                client.delete_object(bucket, obj, options: request_options)
              end
            end
            wait_until { !directory?(path) }
            true
          else
            false
          end
        rescue
          process_error($!)
        end

        def mkdir(path, parents: true, raise_if_exist: false)
          if exist?(path)
            if raise_if_exist
              raise Tumugi::FileAlreadyExistError.new("Path #{path} is already exist")
            elsif !directory?(path)
              raise Tumugi::NotADirectoryError.new("Path #{path} is not a directory")
            end
            false
          else
            put_string("", add_path_delimiter(path))
            true
          end
        rescue
          process_error($!)
        end

        def directory?(path)
          bucket, key = path_to_bucket_and_key(path)
          if root?(key)
            bucket_exist?(bucket)
          else
            obj = add_path_delimiter(key)
            if obj_exist?(bucket, obj)
              true
            else
              # Any objects with this prefix
              objects = client.list_objects(bucket, prefix: obj, max_results: 20, options: request_options)
              !!(objects.items && objects.items.size > 0)
            end
          end
        rescue
          process_error($!)
        end

        def entries(path)
          bucket, key = path_to_bucket_and_key(path)
          obj = add_path_delimiter(key)
          results = []
          next_page_token = ''

          until next_page_token.nil?
            objects = client.list_objects(bucket, prefix: obj, page_token: next_page_token, options: request_options)
            if objects && objects.items
              results.concat(objects.items)
              next_page_token = objects.next_page_token
            else
              next_page_token = nil
            end
          end
          results
        rescue
          process_error($!)
        end

        def move(src_path, dest_path, raise_if_exist: false)
          copy(src_path, dest_path, raise_if_exist: raise_if_exist)
          remove(src_path)
        end

        #######################################################################
        # Specific methods
        #######################################################################

        def upload(media, path, content_type: nil)
          bucket, key = path_to_bucket_and_key(path)
          obj = Google::Apis::StorageV1::Object.new(bucket: bucket, name: key)
          client.insert_object(bucket, obj, upload_source: media, content_type: content_type, options: request_options)
          wait_until { obj_exist?(bucket, key) }
        rescue
          process_error($!)
        end

        def download(path, download_path: nil, mode: 'r', &block)
          bucket, key = path_to_bucket_and_key(path)
          if download_path.nil?
            download_path = Tempfile.new('tumugi_gcs_file_system').path
          end
          client.get_object(bucket, key, download_dest: download_path, options: request_options)
          wait_until { File.exist?(download_path) }

          if block_given?
            File.open(download_path, mode, &block)
          else
            File.open(download_path, mode)
          end
        rescue
          process_error($!)
        end

        def put_string(contents, path, content_type: 'text/plain')
          media = StringIO.new(contents)
          upload(media, path, content_type: content_type)
        end

        def copy(src_path, dest_path, raise_if_exist: false)
          if raise_if_exist && exist?(dest_path)
            raise Tumugi::FileAlreadyExistError.new("Path #{dest_path} is already exist")
          end

          src_bucket, src_key = path_to_bucket_and_key(src_path)
          dest_bucket, dest_key = path_to_bucket_and_key(dest_path)

          if directory?(src_path)
            src_prefix = add_path_delimiter(src_key)
            dest_prefix = add_path_delimiter(dest_key)

            src_path = add_path_delimiter(src_path)
            copied_objs = []
            entries(src_path).each do |entry|
              suffix = entry.name[src_prefix.length..-1]
              client.copy_object(src_bucket, src_prefix + suffix,
                                  dest_bucket, dest_prefix + suffix, options: request_options)
              copied_objs << (dest_prefix + suffix)
            end
            wait_until { copied_objs.all? {|obj| obj_exist?(dest_bucket, obj)} }
          else
            client.copy_object(src_bucket, src_key, dest_bucket, dest_key, options: request_options)
            wait_until { obj_exist?(dest_bucket, dest_key) }
          end
        rescue
          process_error($!)
        end

        def path_to_bucket_and_key(path)
          uri = URI.parse(path)
          raise Tumugi::FileSystemError.new("URI scheme must be 'gs' but '#{uri.scheme}'") unless uri.scheme == 'gs'
          [ uri.host, uri.path[1..-1] ]
        end

        def create_bucket(bucket)
          unless bucket_exist?(bucket)
            b = Google::Apis::StorageV1::Bucket.new(name: bucket)
            client.insert_bucket(@project_id, b, options: request_options)
            true
          else
            false
          end
        rescue
          process_error($!)
        end

        def remove_bucket(bucket)
          if bucket_exist?(bucket)
            client.delete_bucket(bucket, options: request_options)
            true
          else
            false
          end
        rescue
          process_error($!)
        end

        def bucket_exist?(bucket)
          client.get_bucket(bucket, options: request_options)
          true
        rescue => e
          return false if e.status_code == 404
          process_error(e)
        end

        private

        def obj_exist?(bucket, key)
          client.get_object(bucket, key, options: request_options)
          true
        rescue => e
          return false if e.status_code == 404
          process_error(e)
        end

        def root?(key)
          key.nil? || key == ''
        end

        def add_path_delimiter(key)
          if key.end_with?('/')
            key
          else
            "#{key}/"
          end
        end

        def save_config(config)
          if config.private_key_file.nil?
            @project_id = config.project_id
            client_email = config.client_email
            private_key = config.private_key
          else
            json = JSON.parse(File.read(config.private_key_file))
            @project_id = json['project_id']
            client_email = json['client_email']
            private_key = json['private_key']
          end
          @key = {
            client_email: client_email,
            private_key: private_key
          }
        end

        def client
          return @cached_client if @cached_client && @cached_client_expiration > Time.now

          client = Google::Apis::StorageV1::StorageService.new
          scope = Google::Apis::StorageV1::AUTH_DEVSTORAGE_READ_WRITE

          if @key[:client_email] and @key[:private_key]
            options = {
              json_key_io: StringIO.new(JSON.generate(@key)),
              scope: scope
            }
            auth = Google::Auth::ServiceAccountCredentials.make_creds(options)
          else
            auth = Google::Auth.get_application_default([scope])
          end
          auth.fetch_access_token!
          client.authorization = auth

          @cached_client_expiration = Time.now + (auth.expires_in / 2)
          @cached_client = client
        end

        def request_options
          {
            retries: 5,
            timeout_sec: 60
          }
        end

        def wait_until(&block)
          while not block.call
            sleep 1
          end
        end

        def process_error(err)
          if err.respond_to?(:body)
            begin
              jobj = JSON.parse(err.body)
              error = jobj["error"]
              reason = error["errors"].map{|e| e["reason"]}.join(",")
              errors = error["errors"].map{|e| e["message"] }.join("\n")
            rescue JSON::ParserError
              reason = err.status_code.to_s
              errors = "HTTP Status: #{err.status_code}\nHeaders: #{err.header.inspect}\nBody:\n#{err.body}"
            end
            raise Tumugi::FileSystemError.new(errors, reason)
          else
            raise err
          end
        end
      end
    end
  end
end
